Gems/Archive/Code/Source/Clients/ArchiveReader.cpp

```diff
@@ -0,0 +1,885 @@
+                    " does not exist in the archive.", AZ_PATH_ARG(errorResult.m_relativeFilePath)));
+            return errorResult;
+        }
+
+        // Now that the file has been found, pass in the ArchiveFileToken to the
+        // the other overload
+        return ListFileInArchive(static_cast<ArchiveFileToken>(foundIt->second));
+    }
+
+    bool ArchiveReader::ContainsFile(AZ::IO::PathView relativePath) const
+    {
+        return static_cast<bool>(ListFileInArchive(relativePath));
+    }
+
+    EnumerateArchiveResult ArchiveReader::EnumerateFilesInArchive(ListFileCallback listFileCallback) const
+    {
+        // Lambda is marked mutable to allow the filePathIndex variable to be incremented each call
+        auto EnumerateAllFiles = [&listFileCallback, this, filePathIndex = 0]
+        (AZ::u32 filePathBlobOffset, AZ::u16 filePathSize) mutable
+            {
+                // Invoke callback on each file with a non-empty path
+                if (AZ::IO::PathView contentPathView(m_archiveToc.m_tocView.m_filePathBlob.substr(filePathBlobOffset, filePathSize));
+                    !contentPathView.empty())
+                {
+                    ArchiveListFileResult listResult;
+                    listResult.m_relativeFilePath = contentPathView;
+                    listResult.m_filePathToken = static_cast<ArchiveFileToken>(filePathIndex);
+
+                    // Gather the file metadata
+                    const ArchiveTocFileMetadata& fileMetadata = m_archiveToc.m_tocView.m_fileMetadataTable[filePathIndex];
+
+                    // Use the compression algorithm index to lookup the compression algorithm ID
+                    // if the file the value is less than the size of the compression AlgorithmIds array
+                    if (fileMetadata.m_compressionAlgoIndex < m_archiveHeader.m_compressionAlgorithmsIds.size())
+                    {
+                        listResult.m_compressionAlgorithm = m_archiveHeader.m_compressionAlgorithmsIds[
+                            fileMetadata.m_compressionAlgoIndex];
+                    }
+
+                    listResult.m_uncompressedSize = fileMetadata.m_uncompressedSize;
+                    listResult.m_compressedSize = fileMetadata.m_compressedSizeInSectors * ArchiveDefaultBlockAlignment;
+                    listResult.m_offset = fileMetadata.m_offset;
+
+                    listFileCallback(AZStd::move(listResult));
+                }
+                ++filePathIndex;
+            };
+        EnumerateFilePathIndexOffsets(EnumerateAllFiles, m_archiveToc.m_tocView);
+
+        // There are currently no error messages that enumerate file path sets
+        // So a default constructed instance which converts to boolean true is returned
+        return {};
+    }
+
+    bool ArchiveReader::WriteArchiveMetadata(AZ::IO::GenericStream& metadataStream,
+        const ArchiveMetadataSettings& metadataSettings) const
```

Comment:
It seems odd that the reader is writing data 😄

Reply:
Well this is for writing out Archive file informational data for diagnostics.
So if the user wanted to just write out all the file names within the archive in a machine parseable format